src/NHibernate/Mapping/Constraint.cs

```diff
@@ -38,6 +39,142 @@ public IEnumerable<Column> ColumnIterator
+		/// Static so the name can be generated prior to creating the Constraint.
+		/// They're cached, keyed by name, in multiple locations.
+		/// </summary>
+		/// <param name="prefix">A name prefix for the generated name.</param>
+		/// <param name="table">The table for which the name is generated.</param>
+		/// <param name="referencedTable">The referenced table, if any.</param>
+		/// <param name="columns">The columns for which the name is generated.</param>
+		/// <returns>The generated name.</returns>
+		/// <remarks>Hybrid of Hibernate <c>Constraint.generateName</c> and
+		/// <c>NamingHelper.generateHashedFkName</c>.</remarks>
+		public static string GenerateName(
+			string prefix, Table table, Table referencedTable, IEnumerable<Column> columns)
+		{
+			// Use a concatenation that guarantees uniqueness, even if identical names
+			// exist between all table and column identifiers.
+			var sb = new StringBuilder("table`" + table.Name + "`");
+			if (referencedTable != null)
+				sb.Append("references`" + referencedTable.Name + "`");
+
+			// Ensure a consistent ordering of columns, regardless of the order
+			// they were bound.
+			// Clone the list, as sometimes a set of order-dependent Column
+			// bindings are given.
+			var alphabeticalColumns = new List<Column>(columns);
+			alphabeticalColumns.Sort(ColumnComparator.Instance);
+			foreach (var column in alphabeticalColumns)
+			{
+				var columnName = column == null ? "" : column.Name;
+				sb.Append("column`").Append(columnName).Append("`");
+			}
+			// Hash the generated name for avoiding collisions with user choosen names.
+			// This is not 100% reliable, as hashing may still have a chance of generating
+			// collisions.
+			var name = prefix + HashName(sb.ToString());
+
+			// Hibernate uses an algorithm yielding names shorter than 30 characters. But we cannot
+			// use it (see HashName). And also we have DB limited to even less (Informix)...
+			if (name.Length > 30)
+			{
+				// This, of course, increases the collision risk.
+				name = name.Substring(0, 30);
+			}
+
+			return name;
+		}
+
+		#region Name generation support methods
+
+		/// <summary>
+		/// Hash a constraint name. Convert the hash digest to base 32
+		/// (full alphanumeric) for shortening the hash string representation
+		/// while keeping it suitable for db names.
+		/// </summary>
+		/// <param name="name">The name to be hashed.</param>
+		/// <returns>The hased name.</returns>
+		private static string HashName(string name)
+		{
+			// Hibernate uses MD5, but with .Net this would throw on FIPS enabled machine.
+			// As a consequence generated names will be quite longer.
+			using (var hasher = SHA256.Create())
+			{
+				var hash = hasher.ComputeHash(Encoding.UTF8.GetBytes(name));
+				// Converting to base 32 for shortening the name.
+				// Hibernate uses base 35, but we do not have a native implementation
+				// in .Net, and base 32 is easier to implement.
+				return ToBase32String(hash);
```

Comment:
Hibernate creates a `BigInteger` from the byte array (.Net can do that too) then asks its base 35 representation: .Net does not have this natively, and implementing it was a bit heavy in my opinion.